File: realtime-macos/Sources/PAPAluLive/ThoughtCloudPlan.swift
struct ThoughtCloudFrame: Equatable {
    let x: Double
    let y: Double
    let width: Double
    let height: Double
}

struct ThoughtCloudConfiguration: Equatable {
    static let `default` = ThoughtCloudConfiguration(
        appearanceDelay: 0.5,
        fadeDuration: 0.18,
        dotStepInterval: 0.3,
        inactiveDotAlpha: 0.35,
        activeDotAlpha: 1.0,
        normalizedFrame: ThoughtCloudFrame(
            x: 0.525,
            y: 0.725,
            width: 0.45,
            height: 0.27
        )
    )

    let appearanceDelay: Double
    let fadeDuration: Double
    let dotStepInterval: Double
    let inactiveDotAlpha: Double
    let activeDotAlpha: Double
    let normalizedFrame: ThoughtCloudFrame
}

struct ThoughtCloudPlan {
    let configuration: ThoughtCloudConfiguration

    init(configuration: ThoughtCloudConfiguration = .default) {
        self.configuration = configuration
    }

    func nextDotIndex(after currentIndex: Int) -> Int {
        (max(0, currentIndex) + 1) % 3
    }

    func dotAlphas(activeIndex: Int) -> [Double] {
        (0..<3).map { index in
            index == activeIndex
                ? configuration.activeDotAlpha
                : configuration.inactiveDotAlpha
        }
    }

    func frame(windowWidth: Double, windowHeight: Double) -> ThoughtCloudFrame {
        let normalized = configuration.normalizedFrame
        return ThoughtCloudFrame(
            x: normalized.x * windowWidth,
            y: normalized.y * windowHeight,
            width: normalized.width * windowWidth,
            height: normalized.height * windowHeight
        )
    }
}

File: realtime-windows/PAPAluLive.Core/ThoughtCloudPlan.cs
namespace PAPAluLive.Core;

public sealed record ThoughtCloudFrame(
    double X,
    double Y,
    double Width,
    double Height);

public sealed class ThoughtCloudPlan
{
    public const double AppearanceDelay = 0.5;
    public const double FadeDuration = 0.18;
    public const double DotStepInterval = 0.3;
    public const double InactiveDotAlpha = 0.35;
    public const double ActiveDotAlpha = 1.0;

    private static readonly ThoughtCloudFrame NormalizedFrame = new(
        X: 0.525,
        Y: 0.725,
        Width: 0.45,
        Height: 0.27);

    public int NextDotIndex(int currentIndex) => (Math.Max(0, currentIndex) + 1) % 3;

    public IReadOnlyList<double> DotAlphas(int activeIndex) => Enumerable
        .Range(0, 3)
        .Select(index => index == activeIndex ? ActiveDotAlpha : InactiveDotAlpha)
        .ToArray();

    public ThoughtCloudFrame Frame(double windowWidth, double windowHeight) => new(
        X: NormalizedFrame.X * windowWidth,
        Y: NormalizedFrame.Y * windowHeight,
        Width: NormalizedFrame.Width * windowWidth,
        Height: NormalizedFrame.Height * windowHeight);
}
